### PR TITLE
feat: add middleware

### LINKED ISSUE #65: Authorization: create middleware and protect routes.
- [x] only match Next.js routes (not static files)
```
export const config = {
  matcher: [
    /*
     * Match all request paths except for the ones starting with:
     * - _next/static (static files)
     * - _next/image (image optimization files)
     * - favicon.ico (favicon file)
     */
    '/((?!_next/static|_next/image|favicon.ico).*)',
  ],
}
```

- [x] check if `request.url` starts with `dashboard`. If yes, then 
  - [x] call session() from session-server.js and wrap with try catch
  - [x] if nothing wrong, then return NextRespnose.next()
  - [x] if something wrong (catch guard), then console.log() and redirect to ('/login')

### PR DESCRIPTION
**Pull request by rizaardiyanto1412**
For issue: https://github.com/alfonsusac/devscale-eventmaker/issues/65

Add the next js routes on the description create infinite redirect loop. 

So instead I just add dashboard to matcher

### CODE REVIEW

File: src/middleware.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    await jose.jwtVerify(token, secret);
+    return NextResponse.next();
+  } catch (error) {
+    return NextResponse.redirect(new URL("/login", request.url));
+  }
+}
+
+export const config = {
+  matcher: ["/dashboard"],
+};
```
Comment (alfonsusac) [inline src/middleware.js]: We might need to handle cases for other urls too. This is not the preferred solution.
The matcher should match all pages route (not static assets, i.e the one ends with .js, .css).

File: src/middleware.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
+import { NextResponse } from "next/server";
+import { session } from "./lib/server-session";
+import * as jose from "jose";
+
+export async function middleware(request) {
+  const { token } = session();
+  const secretKey = process.env.SECRET_KEY;
+  const secret = new TextEncoder().encode(secretKey);
```
Comment (alfonsusac) [inline src/middleware.js:8]: Why use this if we already have jwt-decode?

File: src/middleware.js [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    return NextResponse.redirect(new URL("/login", request.url));
+  }
+
+  //validating token
+  try {
+    await jose.jwtVerify(token, secret);
+    return NextResponse.next();
+  } catch (error) {
+    return NextResponse.redirect(new URL("/login", request.url));
+  }
```
Comment (alfonsusac) [inline src/middleware.js]: Line 6 - 20 should be in an "If" statement with the condition that `request.url` starts with `/dashboard`

**Review by alfonsusac — CHANGES_REQUESTED** [review]

File: src/middleware.js [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+  const secretKey = process.env.SECRET_KEY;
+  const secret = new TextEncoder().encode(secretKey);
+
+  if (!token) {
+    return NextResponse.redirect(new URL("/login", request.url));
+  }
+
+  //validating token
+  try {
+    await jose.jwtVerify(token, secret);
```
Comment (alfonsusac) [inline src/middleware.js]: Why use this if we already have jwt-decode?

**Review by alfonsusac — CHANGES_REQUESTED** [review]

File: src/middleware.js (hunk 4 above)
Comment (rizaardiyanto1412) [reply src/middleware.js]: @alfonsusac jwt-decode just decode the token, it does not verify if the token valid or not

File: src/middleware.js [hunk 5]
```diff
@@ -0,0 +1,39 @@
+import { NextResponse } from "next/server";
+import { session } from "./lib/server-session";
+import * as jose from "jose";
+
+export async function middleware(request) {
+  //parsing the URL
+  const url = new URL(request.url);
+  const path = url.pathname;
+  //verify token on dashboard URL
+  if (path === "/dashboard") {
```
Comment (alfonsusac) [inline src/middleware.js]: This wont work if the `path` is /dashboard/create

you have to check if the path [__starts with__](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/startsWith) `/dashboard`. Not only equal to

**Review by alfonsusac — CHANGES_REQUESTED** [review]

**Review by alfonsusac — APPROVED** [review]
perfect

**Pull request merged**